{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "508be383-9547-46b8-b5fe-e6280d5f2df6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in c:\\users\\herbi\\anaconda3\\lib\\site-packages (1.84)\n",
      "Requirement already satisfied: pyteomics in c:\\users\\herbi\\anaconda3\\lib\\site-packages (4.7.2)\n",
      "Requirement already satisfied: numpy in c:\\users\\herbi\\anaconda3\\lib\\site-packages (from biopython) (1.26.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython pyteomics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2c1ba67e-9439-4d4a-ae7c-372597255b07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1000 spectra; added protein info to 1000 spectra\n",
      "Processed 2000 spectra; added protein info to 2000 spectra\n",
      "Processed 3000 spectra; added protein info to 3000 spectra\n",
      "Processed 4000 spectra; added protein info to 4000 spectra\n",
      "Processed 5000 spectra; added protein info to 5000 spectra\n",
      "Processed 6000 spectra; added protein info to 6000 spectra\n",
      "Processed 7000 spectra; added protein info to 7000 spectra\n",
      "Processed 8000 spectra; added protein info to 8000 spectra\n",
      "Processed 9000 spectra; added protein info to 9000 spectra\n",
      "Processed 10000 spectra; added protein info to 10000 spectra\n",
      "Processed 11000 spectra; added protein info to 11000 spectra\n",
      "Processed 12000 spectra; added protein info to 12000 spectra\n",
      "Processed 13000 spectra; added protein info to 13000 spectra\n",
      "Processed 14000 spectra; added protein info to 14000 spectra\n",
      "Processed 15000 spectra; added protein info to 15000 spectra\n",
      "Processed 16000 spectra; added protein info to 16000 spectra\n",
      "Processed 17000 spectra; added protein info to 17000 spectra\n",
      "Processed 18000 spectra; added protein info to 18000 spectra\n",
      "Processed 19000 spectra; added protein info to 19000 spectra\n",
      "Processed 20000 spectra; added protein info to 20000 spectra\n",
      "Processed 21000 spectra; added protein info to 21000 spectra\n",
      "Processed 22000 spectra; added protein info to 22000 spectra\n",
      "Processed 23000 spectra; added protein info to 23000 spectra\n",
      "Processed 24000 spectra; added protein info to 24000 spectra\n",
      "Processed 25000 spectra; added protein info to 25000 spectra\n",
      "Processed 26000 spectra; added protein info to 26000 spectra\n",
      "Processed 27000 spectra; added protein info to 27000 spectra\n",
      "Processed 28000 spectra; added protein info to 28000 spectra\n",
      "Processed 29000 spectra; added protein info to 29000 spectra\n",
      "Processed 30000 spectra; added protein info to 30000 spectra\n",
      "Processed 31000 spectra; added protein info to 31000 spectra\n",
      "Processed 32000 spectra; added protein info to 32000 spectra\n",
      "Processed 33000 spectra; added protein info to 33000 spectra\n",
      "Processed 34000 spectra; added protein info to 34000 spectra\n",
      "Processed 35000 spectra; added protein info to 35000 spectra\n",
      "Processed 36000 spectra; added protein info to 36000 spectra\n",
      "Processed 37000 spectra; added protein info to 37000 spectra\n",
      "Processed 38000 spectra; added protein info to 38000 spectra\n",
      "Processed 39000 spectra; added protein info to 39000 spectra\n",
      "Processed 40000 spectra; added protein info to 40000 spectra\n",
      "Processed 41000 spectra; added protein info to 41000 spectra\n",
      "Processed 42000 spectra; added protein info to 42000 spectra\n",
      "Processed 43000 spectra; added protein info to 43000 spectra\n",
      "Processed 44000 spectra; added protein info to 44000 spectra\n",
      "Processed 45000 spectra; added protein info to 45000 spectra\n",
      "Processed 46000 spectra; added protein info to 46000 spectra\n",
      "Processed 47000 spectra; added protein info to 47000 spectra\n",
      "Processed 48000 spectra; added protein info to 48000 spectra\n",
      "Processed 49000 spectra; added protein info to 49000 spectra\n",
      "Processed 50000 spectra; added protein info to 50000 spectra\n",
      "Processed 51000 spectra; added protein info to 51000 spectra\n",
      "Processed 52000 spectra; added protein info to 52000 spectra\n",
      "Processed 53000 spectra; added protein info to 53000 spectra\n",
      "Processed 54000 spectra; added protein info to 54000 spectra\n",
      "Processed 55000 spectra; added protein info to 55000 spectra\n",
      "Processed 56000 spectra; added protein info to 56000 spectra\n",
      "Processed 57000 spectra; added protein info to 57000 spectra\n",
      "Processed 58000 spectra; added protein info to 58000 spectra\n",
      "Processed 59000 spectra; added protein info to 59000 spectra\n",
      "Processed 60000 spectra; added protein info to 60000 spectra\n",
      "Processed 61000 spectra; added protein info to 61000 spectra\n",
      "Processed 62000 spectra; added protein info to 62000 spectra\n",
      "Processed 63000 spectra; added protein info to 63000 spectra\n",
      "Processed 64000 spectra; added protein info to 64000 spectra\n",
      "Processed 65000 spectra; added protein info to 65000 spectra\n",
      "Processed 66000 spectra; added protein info to 66000 spectra\n",
      "Processed 67000 spectra; added protein info to 67000 spectra\n",
      "Processed 68000 spectra; added protein info to 68000 spectra\n",
      "MGF file with added protein information has been saved to: F:\\Libfile\\Celegans_1_addprot\n"
     ]
    }
   ],
   "source": [
    "from bisect import bisect_left\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "from pyteomics import mgf, mass\n",
    "\n",
    "def return_frag_mzs(peptide, z):\n",
    "    mzValues = []\n",
    "    digPat = r'[\\+\\-]\\d+(\\.\\d+)?'  # Updated regex pattern\n",
    "    digs = re.findall(digPat, peptide)\n",
    "    pepFrags = re.split(digPat, peptide)\n",
    "    modValues = {}\n",
    "    seq = ''\n",
    "    while len(digs) != 0:\n",
    "        dig = digs.pop(0)\n",
    "        frag = pepFrags.pop(0)\n",
    "        seq += frag\n",
    "        modValue = float(dig) / z  # Include the sign in conversion\n",
    "        modValues[len(seq)] = modValue\n",
    "    seq += pepFrags[0]\n",
    "    # Calculate y-ions\n",
    "    for i in range(1, len(seq)):\n",
    "        mz = mass.fast_mass(sequence=seq[i:], ion_type='y', charge=z)\n",
    "        mz += sum([modValues[pos] for pos in modValues if pos > i])\n",
    "        mzValues.append(mz)\n",
    "    # Calculate b-ions\n",
    "    for i in range(1, len(seq)):\n",
    "        mz = mass.fast_mass(sequence=seq[:i], ion_type='b', charge=z)\n",
    "        mz += sum([modValues[pos] for pos in modValues if pos <= i])\n",
    "        mzValues.append(mz)\n",
    "    return mzValues\n",
    "\n",
    "def approx_list(value, lst, tolerance=0.5):\n",
    "    for idx, item in enumerate(lst):\n",
    "        if abs(value - item) <= tolerance:\n",
    "            return idx\n",
    "    return -1\n",
    "\n",
    "def clean_mgf_file(mgfFile, fasta, output_file, ions=False):\n",
    "    spectra = mgf.read(mgfFile)\n",
    "    fDict = {}\n",
    "    longPep = ''\n",
    "    \n",
    "    # Parse the FASTA file and create a dictionary of protein sequences\n",
    "    for record in SeqIO.parse(open(fasta, 'r'), 'fasta'):\n",
    "        fDict[len(longPep)] = record.id\n",
    "        longPep += str(record.seq) + '.'\n",
    "    \n",
    "    cleaned = []\n",
    "    count = 0\n",
    "    pepCount = 0\n",
    "    \n",
    "    # Updated regex pattern for modification removal\n",
    "    mod_removal_pattern = r'[\\+\\-]\\d+(\\.\\d+)?'\n",
    "    \n",
    "    for spec in spectra:\n",
    "        count += 1\n",
    "        \n",
    "        if ions:\n",
    "            mzValues = return_frag_mzs(spec['params']['seq'], 1)\n",
    "            peaks = list(zip(spec['m/z array'], spec['intensity array']))\n",
    "            for i in range(len(peaks)-1, -1, -1):\n",
    "                if approx_list(peaks[i][0], mzValues) == -1:\n",
    "                    peaks.pop(i)\n",
    "            if len(peaks) == 0:\n",
    "                continue\n",
    "            peaks.sort(key=lambda x: x[0])\n",
    "            spec['m/z array'], spec['intensity array'] = map(list, zip(*peaks))\n",
    "        \n",
    "        decoy = False\n",
    "        if 'protein' in spec['params'] and 'DECOY' in spec['params']['protein']:\n",
    "            decoy = True\n",
    "        else:\n",
    "            # Remove both positive and negative modifications\n",
    "            seq = re.sub(mod_removal_pattern, '', spec['params']['seq'])\n",
    "            listOfI = [m.start() for m in re.finditer(seq, longPep)]\n",
    "            sorted_keys = sorted(fDict.keys())\n",
    "            proteins = set()\n",
    "            for i in listOfI:\n",
    "                insertion_point = bisect_left(sorted_keys, i)\n",
    "                if insertion_point == len(sorted_keys) or sorted_keys[insertion_point] != i:\n",
    "                    insertion_point -= 1\n",
    "                protein = fDict[sorted_keys[insertion_point]]\n",
    "                proteins.add(protein)\n",
    "            if len(proteins) == 0:\n",
    "                proteins.add('protein_not_in_fasta_' + spec['params']['seq'])\n",
    "\n",
    "        if decoy:\n",
    "            proteins = ['DECOY_0_' + x for x in proteins]\n",
    "\n",
    "        protein = str(len(proteins)) + '/' + '/'.join(sorted(proteins))\n",
    "        if protein != '0/':\n",
    "            spec['params']['protein'] = protein\n",
    "            pepCount += 1\n",
    "        \n",
    "        cleaned.append(spec)\n",
    "        \n",
    "        if count % 1000 == 0:\n",
    "            print(f\"Processed {count} spectra; added protein info to {pepCount} spectra\")\n",
    "    \n",
    "    # Write the cleaned spectra with protein information to a new MGF file\n",
    "    mgf.write(cleaned, output_file)\n",
    "    print(f\"MGF file with added protein information has been saved to: {output_file}\")\n",
    "\n",
    "# Usage example (replace with your actual file paths)\n",
    "mgf_file = r\"F:\\Libfile\\Celegans_1_addseq\"\n",
    "fasta_file = r\"F:\\Libfile\\uniprotkb_proteome_UP000001940_2025_01_23.fasta\"\n",
    "output_file = r\"F:\\Libfile\\Celegans_1_addprot\"\n",
    "\n",
    "clean_mgf_file(mgf_file, fasta_file, output_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "385b168b-a432-48b6-9c9e-b15d51c0aed9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total spectra processed: 27643\n",
      "Total decoy spectra modified: 462\n",
      "Modified MGF file saved as: D:\\DirectInfusion\\Anti-body Monoclonal\\RAW file for Lib\\NIST_mAb_PTM_adddecoysfixed.mgf\n"
     ]
    }
   ],
   "source": [
    "from pyteomics import mgf\n",
    "\n",
    "def add_decoy_to_title(input_mgf, output_mgf):\n",
    "    spectra = mgf.read(input_mgf)\n",
    "    modified_spectra = []\n",
    "    count = 0\n",
    "    decoy_count = 0\n",
    "\n",
    "    for spec in spectra:\n",
    "        count += 1\n",
    "        params = spec['params']\n",
    "        protein = params.get('PROTEIN', params.get('protein', ''))\n",
    "\n",
    "        # Check if 'DECOY' is in the 'PROTEIN' parameter\n",
    "        if 'DECOY' in protein:\n",
    "            decoy_count += 1\n",
    "            # Modify the 'TITLE' parameter\n",
    "            if 'TITLE' in params:\n",
    "                params['TITLE'] = 'DECOY_' + params['TITLE']\n",
    "            elif 'title' in params:\n",
    "                params['title'] = 'DECOY_' + params['title']\n",
    "            else:\n",
    "                # If 'TITLE' doesn't exist, create it\n",
    "                params['TITLE'] = 'DECOY_Spectrum_' + str(count)\n",
    "        else:\n",
    "            # Ensure non-decoy spectra have a 'TITLE' parameter\n",
    "            if 'TITLE' not in params and 'title' not in params:\n",
    "                params['TITLE'] = 'Spectrum_' + str(count)\n",
    "\n",
    "        modified_spectra.append(spec)\n",
    "\n",
    "    # Write the modified spectra to a new MGF file\n",
    "    mgf.write(modified_spectra, output_mgf)\n",
    "\n",
    "    print(f\"Total spectra processed: {count}\")\n",
    "    print(f\"Total decoy spectra modified: {decoy_count}\")\n",
    "    print(f\"Modified MGF file saved as: {output_mgf}\")\n",
    "\n",
    "# Usage example (replace with your actual file paths)\n",
    "input_mgf = r\"D:\\DirectInfusion\\Anti-body Monoclonal\\RAW file for Lib\\NIST_mAb_PTM_adddecoys.mgf\"\n",
    "output_mgf = r\"D:\\DirectInfusion\\Anti-body Monoclonal\\RAW file for Lib\\NIST_mAb_PTM_adddecoysfixed.mgf\"\n",
    "\n",
    "add_decoy_to_title(input_mgf, output_mgf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dbef6e1-d4f3-4bf1-b736-dddeaae020e8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
